[Mono] Added Point classes, Modulus ops, comments, misc C# improvements

aaronfranke opened the pull request:
[See here for the reason why this PR is closed](https://github.com/godotengine/godot/pull/17766#issuecomment-379523741).

<details>

~~I've added new classes called Point2 and Point3. These are similar to Vector2 and Vector3 but work with integers.~~ They are useful in different situations compared to Vectors, one does not replace the other. ~~I've also added Modulus functions to Mathf and to those classes.~~

Example use case for these: Let's say you're creating a game like Minecraft. In Minecraft, the world is divided into 16 by 16 block chunks. You can get the player's block-position within the chunk like this:

    Point3 chunkPos = position.Floor().ToPoint() & 15; // only useful for power-of-two sizes

Or, if you have chunks that are more tall than wide:

    Point3 modPoint = new Point3(15, 255, 15);
    Point3 chunkPos = position.Floor().ToPoint() & modPoint;

If you do want decimals, you can also get the position within the chunk, including decimals, like this:

    Vector3 chunkPos = position % 16; // any number, no restrictions unlike &

`%` is available for Vectors and Points, while `&` is vastly faster than `%` but unique to Points and only available in certain situations due to its bitwise nature. 

Unlike with Vectors, you can directly use the integer values of Points with things like arrays and dictionaries, with exact matching (with Vectors, 2.000001 would be different from 1.999999). These Point classes and methods are useful for any voxel game, or really any game working with integer coordinates or grids.

~~My main concern with this feature is that it's only in C#. I would greatly appreciate it if someone could implement these Point classes in C++/VS/GDscript.~~

~~There are a few other misc improvements in this PR. I've added `<summary>` to many classes that benefit from explanations. I've added more Logarithm types, most of which are wrappers for existing `System.Math` functions. I added a few more constants. I replaced all tabs with spaces, too.~~

Zylann commented:
Point2i and Point3i have been requested in the scripting API a year ago https://github.com/godotengine/godot/issues/3286

NathanWarden commented:
I can definitely see a benefit in having ints instead of floats, thanks Aaron.

vnen commented:
I understand the usefulness of Point2/3 types, but having them exclusive to C# makes it diverges from the engine API, which should not really happen. If it's really useful, should be available to everyone, not only C# users.

aaronfranke commented:
@vnen I agree completely, that's why I wrote:

> My main concern with this feature is that it's only in C#. I would greatly appreciate it if someone could implement these Point classes in C++/VS/GDscript.

I am not an expert with C++ and GDscript, nor with most of the Godot code.

maikramer commented:
I have classes to this, called Vector2Int, I use in my procedural generation maps, because they can be used in a for loop, a switch, and other places where only int are allowed. I would like to share my types also, but I'm in the same situation, I'm only confident with c#, and I agree that the classes must be added in cpp then can be used in any scripting language.
At same time, if we could add some things just in mono, people that is familiar with c# only could add lots of features, like, I made extensions to node to have Find<T>() (find node by type and so on.

maikramer commented:
And Aaron, i miss what you did with Point2/3 in vector3 and 2, like Vector3.Zero and One property, would be good additions, now I need to do new Vector3(1,1,1) * scale.

maikramer commented:
@aaronfranke thanks, i apreciate this a lot.

aaronfranke commented:
I made a few tweaks to this commit. Most notably, I allowed implicit conversion of Points to Vectors:

    public static implicit operator Vector3(Point3 point) 
    {
        return new Vector3(point);
    }

Now, you have the ability to write code like this:

    Point3 positionPoint = new Point3(5, 6, 7);
    Translation = positionPoint;

This also means that you can use a Point class for literally any method that accepts a Vector! `Translate (Point3.One);`? No problem!

I did not add any implicit conversions from Vector to Point because information is lost (3.3 becomes 3 etc), so if you wish to convert from Vector to Point, you should do it explicitly with `ToPoint()`.

maikramer commented:
Another thing I have by myself that I didn't see in Godot is InverseLerp. I use to normalize my height map in a range 0 to 1. Just a idea for addition, even if we can do by ourselves, what is pretty easy, rely on official versions always make our code simpler. Keep your good work.

akien-mga commented on modules/mono/glue/cs_files/Basis.cs:
```diff
@@ -381,47 +381,47 @@ public Vector3 XformInv(Vector3 v)
-				);
-			}
-		}
+        public Quat Quat() {
```
Please split indentation changes in their own commit (if 4-space is the agreed upon style for C# files), they make the rest of the functional changes unreadable.

aaronfranke committed:
[Mono] Replace tabs with spaces, some descriptions, spacing

[Mono] Replace tabs with spaces, some descriptions, spacing

aaronfranke committed:
[Mono] Added Point classes, Modulus ops, comments, misc C# improvements

[Mono] Added Point2 and Point3 classes, comments to many classes that benefit from explanations, Modulus operations, more Logarithm types, more constants, replace tabs with spaces, more misc C# improvements.

Chaosus commented on modules/mono/glue/cs_files/Mathf.cs:
```diff
@@ -208,6 +321,21 @@ public static real_t Rad2Deg(real_t rad)
+
+        public static int RemInt(int a, int b) 
+        {
+            return (a % b);
```
I dont understand the usefullness of this function, by this logic you must also implement add(int a, int b), subract(int a, int b) etc... its redundant, because its basic operation, and its not a problem for user to write "a % b" himself

aaronfranke replied:
It is useful because `%` is ambiguous and has different meanings in different programming languages. In Python, % is Mod, but in C-derived languages, it's Rem. But as they are equivalent for positive values, programmers may not be aware of the different behavior with negative numbers. `Mathf.Rem()` allows saying "I explicitly want Remainder and not Modulus".

Chaosus commented on modules/mono/glue/cs_files/Mathf.cs:
```diff
@@ -139,28 +167,57 @@ public static real_t Floor(real_t s)
+
+        public static real_t LogE(real_t s)
+        {
+            return (real_t)Math.Log(s);
```
Why you create two functions with same functionality(LogE and Log) ?

aaronfranke replied:
`Log` for backwards compatibility and the fact that it's the same name as in System.Math, but I prefer the term `LogE` as it's more clear and follows the pattern of `Log10` and `Log2`.

Chaosus replied:
And its not be confusable for other users ?

aaronfranke replied:
LogE is a less ambiguous term. People looking for natural log will be able to easily find it with LogE, people looking for just any log or a wrapper for Math.Log will look for Mathf.Log.

Lots of these changes mostly improve the readability of a game's code.

Chaosus commented on modules/mono/glue/cs_files/Mathf.cs:
```diff
@@ -181,6 +238,57 @@ public static real_t Min(real_t a, real_t b)
+        /// </summary>
+        public static int ModBit2(int a, int b)
+        {
+            return (a & b);
```
Its a basic operation between two integers, the user can do it himself, without calling a function(which is not a fast operation)..

aaronfranke replied:
AFAIK C# JIT has an inline optimization for methods that are compiled to 32 bytes or less.

`ModBit2` does the same as `&`, yes, but it states the purpose of the operation in code. I also have this method defined in Point classes so it's nice to have a method in Mathf.

neikeq replied:
Yes, the compiler does a good job at optimizing this. In fact, I was testing the other day with a conversion operator between two Vector3 types:

``` C#
public static implicit operator Godot.Vector3(OpenTK.Vector3 from)
{
    return new Godot.Vector3(from.X, from.Y, from.Z);
}


OpenTK.Vector3 a = new OpenTK.Vector3();
Godot.Vector3 b = a;
```

I was very surprise to see that the compiler optimizes it to be the same as a fast memory copy:

``` C#
Godot.Vector3 a = new Godot.Vector3();
Godot.Vector3 b = a;
```

Chaosus commented on modules/mono/glue/cs_files/Point2.cs:
```diff
@@ -0,0 +1,419 @@
+
+        public Point2 Abs()
+        {
+            return new Point2( Mathf.RoundToInt(Mathf.Abs(x)), Mathf.RoundToInt(Mathf.Abs(y)) );
```
Could be changed to <code>return new Point2( Mathf.Abs(x), Mathf.Abs(y) )</code> if you create Math.Abs for integers.

Chaosus commented:
In my opinion this PR is a mess, I think all changes(except maybe some syntax sugar) must be reasonable and compatible between GDScript and C#

aaronfranke commented on modules/mono/glue/cs_files/Mathf.cs:
```diff
@@ -181,6 +238,57 @@ public static real_t Min(real_t a, real_t b)
+                c += b;
+            }
+            return c;
+        }
```
Similarly with Log and LogE, I prefer the term `Mod` but Godot already has `Fposmod`. So I kept it for backwards compatibility but I tweaked it to `Mod` as I've found this approach runs faster.

aaronfranke commented on modules/mono/glue/cs_files/Point3.cs:
```diff
@@ -0,0 +1,471 @@
+
+        public Point3 Abs()
+        {
+            return new Point3( Mathf.RoundToInt(Mathf.Abs(x)), Mathf.RoundToInt(Mathf.Abs(y)), Mathf.RoundToInt(Mathf.Abs(z)) );
```
I agree completely. I'll make this change after we discuss other changes.

aaronfranke commented:
I'm not familiar with the GDScript/C++ part of the Godot codebase. I would appreciate these things being implemented there but I can't do it myself.

Chaosus commented:
I already created cleaner improved version of Mathf in #18038. Dont know what to do now, I hope @neikeq  and @akien-mga can found the solution

aaronfranke commented on modules/mono/glue/cs_files/Point3.cs:
```diff
@@ -0,0 +1,471 @@
+    /// It is useful, for example, when working with points on a 3D grid.
+    /// Not all concepts of Vector3 apply, such as normalization and any functions that depend on normalization. 
+    /// However, new concepts can exist with Point3 such as bitwise operations.
+    /// </summary>
```
Perhaps it is too long. I will probably remove the last two lines, as they aren't really necessary.

aaronfranke commented:
@maikramer, just so [you know](https://github.com/godotengine/godot/pull/17766#issuecomment-377534121), @Chaosus has added `InverseLerp` here: https://github.com/godotengine/godot/pull/18038

aaronfranke commented:
Ok, I'm going to close this PR for now. Thanks for reviewing @Chaosus I will keep these things in mind.

I will re-create it after https://github.com/godotengine/godot/pull/18038 and https://github.com/godotengine/godot/pull/18053 are fixed up & merged, and we further discuss C# styles: https://github.com/godotengine/godot-docs/issues/1296

I'm also going to split it up into several smaller changes. It seems like large PRs are annoying for everyone, so I'll try to split future changes up into bite-sized single-purpose PRs.

aaronfranke commented:
[GridMap](https://godot.readthedocs.io/en/3.0/classes/class_gridmap.html) would benefit from Point3.

maikramer commented:
@aaronfranke yes i saw, my favorite task is look the commits every hour, i'm comming from unity and i'm sure i'm not the only one, csharp is really powerful and i dont see myself using gdscript, its good but not as powerful, i'm using csharp with newtonsoft.json and many other things that make you less tight to the engine. I hope you guys will continue with your job, i would like to help but i'm not professional class like you and the others.

aaronfranke commented:
@maikramer You don't need professional experience to contribute to Godot. All my commits so far have been following the patterns of existing code, and (something anyone can do) discussing changes.

Personally, I have never been paid to write code before. All my C# experience is with hobby projects, a lot of which was with the Unity game engine. Yet I can still contribute :)
